src/audio/mixin_mixout.c

```diff
@@ -885,103 +1086,130 @@ static int mixinout_trigger(struct comp_dev *dev, int cmd)
+				IPC4_MIXER_CHANNEL_REMAPPING_MODE)
+				sink_c->stream.channels =
+					md->sink_config[sink_id].output_channel_count;
+		} else {
+			comp_err(dev, "Sink index out of range: %u, max sink count: %u",
+				 (uint32_t)sink_id, MIXIN_MAX_SINKS);
+			buffer_release(sink_c);
+			return -EINVAL;
+		}
 
-		/* 8 ch stream is supported by ch_map and each channel
-		 * is mapped by 4 bits. The first channel will be mapped
-		 * by the bits of 0 ~ 3 and the 2th channel will be mapped
-		 * by bits 4 ~ 7. The N channel will be mapped by bits
-		 * N*4 ~ N*4 + 3
+		/* comp_verify_params() does not modify valid_sample_fmt (a BUG?),
+		 * let's do this here
 		 */
-		for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
-			sink_c->chmap[i] = (base_cfg->audio_fmt.ch_map >> i * 4) & 0xf;
+		audio_stream_fmt_conversion(md->base_cfg.audio_fmt.depth,
+					    md->base_cfg.audio_fmt.valid_bit_depth,
+					    &sink_c->stream.frame_fmt,
+					    &sink_c->stream.valid_sample_fmt,
+					    md->base_cfg.audio_fmt.s_type);
 
 		buffer_release(sink_c);
 	}
 
+	/* use BUFF_PARAMS_CHANNELS to skip updating channel count */
 	ret = comp_verify_params(dev, BUFF_PARAMS_CHANNELS, params);
 	if (ret < 0) {
-		comp_err(dev, "mixinout_common_params(): comp_verify_params() failed!");
+		comp_err(dev, "mixin_params(): comp_verify_params() failed!");
 		return -EINVAL;
 	}
 
 	return 0;
 }
 
-static int mixin_params(struct comp_dev *dev,
-			struct sof_ipc_stream_params *params)
-{
-	struct mixin_data *md;
-
-	comp_dbg(dev, "mixin_params()");
-
-	md = comp_get_drvdata(dev);
-	return mixinout_common_params(dev, params, &md->base_cfg);
-}
-
 static int mixout_params(struct comp_dev *dev,
 			 struct sof_ipc_stream_params *params)
 {
 	struct mixout_data *md;
-	int ret;
-	struct comp_buffer *sinkb;
+	struct comp_buffer *sink;
 	struct comp_buffer __sparse_cache *sink_c;
+	int ret;
 	uint32_t sink_period_bytes, sink_stream_size;
 
 	comp_dbg(dev, "mixout_params()");
 
 	md = comp_get_drvdata(dev);
+	base_module_cfg_to_stream_params(&md->base_cfg, params);
 
-	ret = mixinout_common_params(dev, params, &md->base_cfg);
-	if (ret < 0)
-		return ret;
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       source_list);
+	sink_c = buffer_acquire(sink);
+
+	/* comp_verify_params() does not modify valid_sample_fmt (a BUG?), let's do this here */
+	audio_stream_fmt_conversion(md->base_cfg.audio_fmt.depth,
+				    md->base_cfg.audio_fmt.valid_bit_depth,
+				    &sink_c->stream.frame_fmt,
+				    &sink_c->stream.valid_sample_fmt,
+				    md->base_cfg.audio_fmt.s_type);
+
+	buffer_release(sink_c);
 
-	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
-				source_list);
-	sink_c = buffer_acquire(sinkb);
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "mixout_params(): comp_verify_params() failed!");
+		return -EINVAL;
+	}
+
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer,
+			       source_list);
```

Comment:
is it now going to be a different buffer? You probably cannot call `comp_verify_params()` while holding a buffer, because it tries to acquire buffers too? But at least you probably can reuse `sink`

Reply:
Modified to reuse sink.